File: java_impl/RandomSupport.java
package com.nyaa.common.util.random;

public class RandomSupport {
    public static final String BAD_SIZE = "size must be non-negative";
    public static final String BAD_DISTANCE = "jump distance must be finite, positive, and an exact integer";
    public static final String BAD_BOUND = "bound must be positive";
    public static final String BAD_FLOATING_BOUND = "bound must be finite and positive";
    public static final String BAD_RANGE = "bound must be greater than origin";

    public static void checkStreamSize(long streamSize) {
        if (streamSize < 0L) {
            throw new IllegalArgumentException(BAD_SIZE);
        }
    }

    public static void checkBound(float bound) {
        if (!(0.0f < bound && bound < Float.POSITIVE_INFINITY)) {
            throw new IllegalArgumentException(BAD_FLOATING_BOUND);
        }
    }

    public static void checkBound(double bound) {
        if (!(0.0 < bound && bound < Double.POSITIVE_INFINITY)) {
            throw new IllegalArgumentException(BAD_FLOATING_BOUND);
        }
    }

    public static void checkBound(int bound) {
        if (bound <= 0) {
            throw new IllegalArgumentException(BAD_BOUND);
        }
    }

    public static void checkBound(long bound) {
        if (bound <= 0) {
            throw new IllegalArgumentException(BAD_BOUND);
        }
    }

    public static void checkRange(float origin, float bound) {
        if (!(Float.NEGATIVE_INFINITY < origin && origin < bound &&
                bound < Float.POSITIVE_INFINITY)) {
            throw new IllegalArgumentException(BAD_RANGE);
        }
    }

    public static void checkRange(double origin, double bound) {
        if (!(Double.NEGATIVE_INFINITY < origin && origin < bound &&
                bound < Double.POSITIVE_INFINITY)) {
            throw new IllegalArgumentException(BAD_RANGE);
        }
    }

    public static void checkRange(int origin, int bound) {
        if (origin >= bound) {
            throw new IllegalArgumentException(BAD_RANGE);
        }
    }

    public static void checkRange(long origin, long bound) {
        if (origin >= bound) {
            throw new IllegalArgumentException(BAD_RANGE);
        }
    }
}

File: ts_impl/random-support.ts
import { Int64_MAXn } from './data-cast'

class RandomSupport {
    static readonly BAD_SIZE = "size must be non-negative";
    static readonly BAD_DISTANCE = "jump distance must be finite, positive, and an exact integer";
    static readonly BAD_BOUND = "bound must be positive";
    static readonly BAD_FLOATING_BOUND = "bound must be finite and positive";
    static readonly BAD_RANGE = "bound must be greater than origin";

    public static checkStreamSize(streamSize: bigint) {
        if (streamSize < 0n) {
            throw new TypeError(RandomSupport.BAD_SIZE);
        }
    }

    public static checkBoundFloat(bound: number) {
        if (!(0.0 < bound && bound < Number.POSITIVE_INFINITY)) {
            throw new TypeError(RandomSupport.BAD_FLOATING_BOUND);
        }
    }

    public static checkBoundDouble(bound: number) {
        if (!(0.0 < bound && bound < Number.POSITIVE_INFINITY)) {
            throw new TypeError(RandomSupport.BAD_FLOATING_BOUND);
        }
    }

    public static checkBoundInt(bound: number) {
        if (bound <= 0) {
            throw new TypeError(RandomSupport.BAD_BOUND);
        }
    }

    public static checkBoundLong(bound: bigint) {
        if (bound <= 0) {
            throw new TypeError(RandomSupport.BAD_BOUND);
        }
    }

    public static checkBoundULong(bound: bigint) {
        if (bound <= 0 || bound > Int64_MAXn) {
            throw new TypeError(RandomSupport.BAD_BOUND);
        }
    }

    public static checkRangeFloat(origin: number, bound: number) {
        if (!(Number.NEGATIVE_INFINITY < origin && origin < bound &&
            bound < Number.POSITIVE_INFINITY)) {
            throw new TypeError(RandomSupport.BAD_RANGE);
        }
    }

    public static checkRangeDouble(origin: number, bound: number) {
        if (!(Number.NEGATIVE_INFINITY < origin && origin < bound &&
            bound < Number.POSITIVE_INFINITY)) {
            throw new TypeError(RandomSupport.BAD_RANGE);
        }
    }

    public static checkRangeInt(origin: number, bound: number) {
        if (origin >= bound) {
            throw new TypeError(RandomSupport.BAD_RANGE);
        }
    }

    public static checkRangeLong(origin: bigint, bound: bigint) {
        if (origin >= bound) {
            throw new TypeError(RandomSupport.BAD_RANGE);
        }
    }

    public static checkRangeULong(origin: bigint, bound: bigint) {
        if (origin < 0 || origin >= bound) {
            throw new TypeError(RandomSupport.BAD_RANGE);
        }
    }
}

export { RandomSupport };
